exes/ethercat/inc/public/tfc/ec/devices/base.hpp

```diff
@@ -45,15 +45,194 @@ concept mp_units_quantity_setting_c = requires {
+  void foobar(int const&, double const&) {}
+};
+static_assert(std::same_as<first_arg_t<decltype(&Foo::foo)>, int>);
+static_assert(std::same_as<first_arg_t<decltype(&Foo::bar)>, int&>);
+static_assert(std::same_as<first_arg_t<decltype(&Foo::foobar)>, int const&>);
+static_assert(std::same_as<second_arg_t<decltype(&Foo::foo)>, double>);
+static_assert(std::same_as<second_arg_t<decltype(&Foo::bar)>, double&>);
+static_assert(std::same_as<second_arg_t<decltype(&Foo::foobar)>, double const&>);
+
+static_assert(!is_first_arg_const_ref<decltype(&Foo::foo)>);
+static_assert(!is_first_arg_const_ref<decltype(&Foo::bar)>);
+static_assert(is_first_arg_const_ref<decltype(&Foo::foobar)>);
+static_assert(!is_second_arg_ref<decltype(&Foo::foo)>);
+static_assert(is_second_arg_ref<decltype(&Foo::bar)>);
+static_assert(!is_second_arg_ref<decltype(&Foo::foobar)>);
+
+}  // namespace details
+
+template <typename impl_t>
 class base {
 public:
-  virtual ~base();
+  using default_t = std::span<std::uint8_t>;
+  base(base const&) = delete;
+  auto operator=(base const&) -> base& = delete;
+  base(base&&) = default;
+  base& operator=(base&&) = default;
 
   // Default behaviour no data processing
-  virtual void process_data(std::span<std::byte>, std::span<std::byte>) = 0;
+  void process_data(std::span<std::uint8_t> input, std::span<std::uint8_t> output) {
+    static_assert(std::is_member_function_pointer_v<decltype(&impl_t::pdo_cycle)>, "impl_t must have method pdo_cycle");
+    using pdo_cycle_func_t = decltype(&impl_t::pdo_cycle);  // is function pointer
+    using input_pdo = std::decay_t<details::first_arg_t<pdo_cycle_func_t>>;
+    using output_pdo = std::decay_t<details::second_arg_t<pdo_cycle_func_t>>;
+
+    // Verify that the declaration of arguments are correct
+    if constexpr (!std::same_as<input_pdo, default_t>) {
+      static_assert(stx::is_constexpr_default_constructible_v<input_pdo>);
+      static_assert(details::is_first_arg_const_ref<pdo_cycle_func_t>,
+                    "impl_t::pdo_cycle must have first argument as const reference");
+    }
+    if constexpr (!std::same_as<output_pdo, default_t>) {
+      static_assert(stx::is_constexpr_default_constructible_v<output_pdo>);
+      static_assert(details::is_second_arg_ref<pdo_cycle_func_t>,
+                    "impl_t::pdo_cycle must have second argument as reference, const is not allowed");
+    }
+
+    // todo simplify
+
+    // Both input and output are default types
+    if constexpr (std::same_as<input_pdo, default_t> && std::same_as<output_pdo, default_t>) {
+      static_cast<impl_t*>(this)->pdo_cycle(input, output);
+    }
+    // input is default type, output is not
+    else if constexpr (std::same_as<input_pdo, default_t>) {
+      if (output.size() != sizeof(output_pdo)) {
+        if (output_buffer_valid_) {
+          logger_.info("Output size mismatch, expected {} bytes, got {} bytes", sizeof(output_pdo), output.size());
+        }
+        output_buffer_valid_ = false;
```

Comment:
Just more nice to read.
```suggestion
            output_buffer_valid_ = false;
        }
```